File: Problems 0200-0299/274. H-Index/solution.py
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        numPapers = len(citations)
        # Observation: h is capped at numPapers
        buckets = [0]*(numPapers + 1)

        for c in citations:
            buckets[min(c, numPapers)] += 1
        
        count = 0
        for h in range(numPapers, -1, -1):
            count += buckets[h]
            if count >= h:
                return h
        return 0

File: Problems 0200-0299/274. H-Index/solution.java
class Solution {
    public int hIndex(int[] citations) {
        // Observation: h-index is capped at citations.length
        int numPapers = citations.length;
        
        int[] buckets = new int[numPapers + 1];
        for (int numCitations : citations) {
            if (numCitations > numPapers) numCitations = numPapers;
            buckets[numCitations]++;
        }

        int count = 0;
        for (int h = numPapers; h >= 0; h--) {
            count += buckets[h];

            if (count >= h) return h;
        }

        return -1;
    }
}